python/ray/data/_internal/execution/bulk_executor.py

```diff
@@ -0,0 +1,99 @@
+
+import ray
+from ray.data._internal.execution.interfaces import (
+    Executor,
+    ExecutionOptions,
+    RefBundle,
+    PhysicalOperator,
+)
+from ray.data._internal.execution.operators.input_data_buffer import InputDataBuffer
+from ray.data._internal.progress_bar import ProgressBar
+from ray.data._internal.stats import DatasetStats
+
+logger = logging.getLogger(__name__)
+
+
+class BulkExecutor(Executor):
+    def __init__(self, options: ExecutionOptions):
+        super().__init__(options)
+        self._stats = DatasetStats(stages={}, parent=None)
+        self._executed = False
+
+    def execute(
+        self, dag: PhysicalOperator, initial_stats: Optional[DatasetStats] = None
+    ) -> Iterator[RefBundle]:
+        """Synchronously executes the DAG via bottom-up recursive traversal."""
+
+        assert not self._executed, "Can only call execute once."
+        self._executed = True
+        if not isinstance(dag, InputDataBuffer):
+            logger.info("Executing DAG %s", dag)
+
+        if initial_stats:
+            self._stats = initial_stats
+
+        saved_outputs: Dict[PhysicalOperator, List[RefBundle]] = {}
+
+        def execute_recursive(op: PhysicalOperator) -> List[RefBundle]:
+            # Avoid duplicate executions.
+            if op in saved_outputs:
+                return saved_outputs[op]
+
+            # Compute dependencies.
+            inputs = [execute_recursive(dep) for dep in op.input_dependencies]
+
+            # Fully execute this operator.
+            logger.debug("Executing op %s", op.name)
+            builder = self._stats.child_builder(op.name)
+            try:
+                for i, ref_bundles in enumerate(inputs):
+                    for r in ref_bundles:
+                        op.add_input(r, input_index=i)
+                    op.inputs_done(i)
+                output = _naive_run_until_complete(op)
+            finally:
+                op.shutdown()
+
+            # Cache and return output.
+            saved_outputs[op] = output
+            op_stats = op.get_stats()
+            op_metrics = op.get_metrics()
+            if op_stats:
+                self._stats = builder.build_multistage(op_stats)
+                self._stats.extra_metrics = op_metrics
+            return output
+
+        return execute_recursive(dag)
+
+    def get_stats(self) -> DatasetStats:
+        assert self._stats is not None, self._stats
+        return self._stats
+
+
+def _naive_run_until_complete(op: PhysicalOperator) -> List[RefBundle]:
+    """Run this operator until completion, assuming all inputs have been submitted.
+
+    Args:
+        op: The operator to run.
+
+    Returns:
+        The list of output ref bundles for the operator.
+    """
+    output = []
+    tasks = op.get_work_refs()
+    if tasks:
+        bar = ProgressBar(op.name, total=op.num_outputs_total())
+        while tasks:
+            done, _ = ray.wait(tasks, fetch_local=True, timeout=0.1)
```

Comment:
This avoids high-CPU spins. Btw, this isn't new code, just copying from the existing implementation of bulk wait.

Reply:
>This avoids high-CPU spins

Hmm IIRC the `ray.wait()` should result in a condition variable wait somewhere in Ray Core, and shouldn't that not result in a high-CPU spin? Unless we're not compiling Ray with the correct flags on Linux...

>also gives us a chance to interrupt

`ray.wait()` is already interruptible, it checks for signals every second regardless of the user-provided timeout IIRC.

>Btw, this isn't new code, just copying from the existing implementation of bulk wait.

Understood, but we really don't want to cargo cult suboptimal code into the new execution model, if we can help it. We've already decided that we should use this redesign as an opportunity to address some long-standing tech debt and raise our quality bar.

Reply:
This isn't the new execution model though--- just the legacy shim until we write the streaming executor. Hence, I'd like to avoid unnecessary execution detail changes that could cause behavior changes.

I do think this is the correct choice though for a couple reasons:
- Setting timeout=0 has historically caused issues with not returning all futures in a timely manner. I do not know if this issue has been fully addressed.
- Setting timeout=large value is problematic for obvious reasons.
- Setting no timeout is also problematic since then you need to set num_returns, breaking batching.

Hence, this is a reasonable, if conservative, choice.

Reply:
>This isn't the new execution model though--- just the legacy shim until we write the streaming executor. Hence, I'd like to avoid unnecessary execution detail changes that could cause behavior changes.

That makes sense, we can keep to the old implementation for the legacy shim and defer any changes until the streaming executor PR.

>Setting timeout=0 has historically caused issues with not returning all futures in a timely manner. I do not know if this issue has been fully addressed.

I don't think this would be ideal anyways, since we'd just be taking the relatively efficient Core Worker wait + a periodic signal check with an inefficient application-level spin loop. So agreed that this wouldn't be a good route.

> Setting no timeout is also problematic since then you need to set num_returns, breaking batching.

Hmm what do you mean by "batching" here?

If you mean batching of multiple "ready" returns, `num_returns` is set to 1 by default, so that kind of batching is already not happening here since when `num_returns=1`, we immediately return a single `ready` ref when it becomes available, right? Quick example demonstrating these unbatched semantics:
```
In [2]: o1 = ray.put(1)

In [3]: o2 = ray.put(2)

In [4]: done, not_done = ray.wait([o1, o2], fetch_local=True, timeout=0.1)

In [5]: done
Out[5]: [ObjectRef(00ffffffffffffffffffffffffffffffffffffff0100000002000000)]

In [6]: not_done
Out[6]: [ObjectRef(00ffffffffffffffffffffffffffffffffffffff0100000001000000)]
```

If you are talking about concurrent object pulls (which I don't think you are), that should have been fixed in Ray Core (with the bug + fix that you discovered) so `len(tasks)` is fetched instead of `num_returns`. https://github.com/ray-project/ray/pull/30724

In any case, unlike the stream executor which wants to immediately know that a task has finished in order to (potentially) launch new work (and should therefore have no timeout for its wait loop), the bulk executor only cares about updating the user-facing progress and stats reporting, so batching at a 100ms granularity is pretty reasonable. If that 100ms batching is the intention, then I think that we would want to specify `num_returns`:
```suggestion
            done, _ = ray.wait(tasks, num_returns=len(tasks), fetch_local=True, timeout=0.1)
```

But yeah, we can just keep the existing bulk execution wait loop behavior, since this is just a legacy shim.

Reply:
> If you mean batching of multiple "ready" returns, num_returns is set to 1 by default

Ah, good catch. Seems like I missed this, indeed this is the batching I meant.